### PR TITLE
update README files for security components

### PR DESCRIPTION
**Pull request by tingyuz**
Add multiple README files for security services and update README for EdgeX-Go to include information of security services.

Signed-off-by: Tingyu Zeng <tingyu.zeng@rsa.com>

### CODE REVIEW

File: SECURITY.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+1. The user needs to create an access token and associate every REST request with the security token
+while sending the request. Use "admin" as groupname below as it is the default enable group in the configuration of proxy.
+
+```sh
+docker-compose run security-proxy-setup --useradd=<account> --group=<groupname>
+```
+
+which will create an access token. The token will be like this: eyJpc3MiOiI5M3V3cmZBc0xzS2Qwd1JnckVFdlRzQloxSmtYOTRRciIsImFjY291bnQiOiJhZG1pbmlzdHJhdG9yIn0
+
+2. The exported external ports (such as 48080, 48081 etc.) will be inaccessible due to security enhancement. Instead all the REST requests need to go through the proxy, and the proxy will redirect the request to individual microservice on behalf of the user.
```
Comment (mkbhanda) [inline SECURITY.md]: s/due to/due to the
s/microservice/microservices

File: SECURITY.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+E.g, if we need to send a request to the ping endpoint of coredata, the original request be like this below:
```
Comment (mkbhanda) [inline SECURITY.md]: s/be like this below/was as below:

File: SECURITY.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Security Components in EdgeX Foundry Services
+
+Starting from Fuji release, EdgeX includes enhanced security features. There are 3 major components that are responsible for security
+features: 
+
+- [Security-secrets-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secrets-setup/README.md)
+- [Security-secretstore-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secretstore-setup/README.md)
+- [Security-proxy-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-proxy-setup/README.md)
```
Comment (bnevis-i) [inline SECURITY.md]: The space between the right bracket and left paren is causing this to render incorrectly.  Also, would a relative URL to README.md work?

File: SECURITY.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+which will create an access token. The token will be like this: eyJpc3MiOiI5M3V3cmZBc0xzS2Qwd1JnckVFdlRzQloxSmtYOTRRciIsImFjY291bnQiOiJhZG1pbmlzdHJhdG9yIn0
+
+2. The exported external ports (such as 48080, 48081 etc.) will be inaccessible due to security enhancement. Instead all the REST requests need to go through the proxy, and the proxy will redirect the request to individual microservice on behalf of the user.
+E.g, if we need to send a request to the ping endpoint of coredata, the original request be like this below:
+
+```
+curl http://{coredata-microservice-ip}:48080/api/v1/ping
+```
+
+Such request needs to be converted into a request like this below:
```
Comment (mkbhanda) [inline SECURITY.md]: s/Such request needs to be converted into a request like this below:/Such a request is converted as below:

File: README.md [hunk 5]
```diff
@@ -21,6 +21,23 @@ You can check the status of your running EdgeX services by going to http://local
...
+- Security-secretstore-setup
+- Security-proxy-setup
+
+When security features are enabled, there are extra steps required to access the resources of EdgeX.
+
+1. The user needs to create an access token and associate every REST request with the access token
+while sending the request. 
+2. The exported external ports (such as 48080, 48081 etc.) will be inaccessible due to security enhancement. Instead all the REST requests need to go through the proxy, and the proxy will redirect the request to individual microservice on behalf of the user.
+
+A sample of steps to create an access token and use the token to access resources of EdgeX can be found here [Security Components](https://github.com/edgexfoundry/edgex-go/blob/master/SECURITY.md)
```
Comment (bnevis-i) [inline README.md]: Would a relative URL work here [ Security Components ] ( SECURITY.md )?

File: SECURITY.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+```
+curl -k https://{proxy-service-ip}:8443/coredata/api/v1/ping -H "Authorization: Bearer <access-token>"
```
Comment (mkbhanda) [inline SECURITY.md]:
```
Please note in the above the use of the proxy IP address,  its port and the need to specify the authorisation header and  the access token therein.

File: cmd/security-proxy-setup/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
+# EdgeX Foundry Security Service - Security Proxy Setup
+[![license](https://img.shields.io/badge/license-Apache%20v2.0-blue.svg)](LICENSE)
+
+Go implementation of EdgeX security-proxy-setup service.
+
+## Features
+
+- Reverse proxy for the existing edgex microservices
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: s/edgex/EdgeX

File: cmd/security-proxy-setup/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+- Account creation with optional either OAuth2 or JWT authentication for existing services
+- Account creation with arbitrary ACL gourp list
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: s/gourp/group

File: cmd/security-proxy-setup/README.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+## Run security-proxy-setup with different parameters
+
+The binary supports multiple command line parameters 
+
+```sh
+ --init  //run init procedure for security service
+ --insecureSkipVerify //skip server side SSL verification, mainly for self-signed cert
+ --configfile //use different configuration file other than the default
+ --reset //reset proxy by removing all customerization
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: s/customerization/customizations

File: cmd/security-proxy-setup/README.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ --useradd //user that needs to be added to consume the edgex services, need to used with 'group' parameter
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: should we not use EdgeX instead of edgeX?

Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: s/need/needs
s/with/with the

File: cmd/security-proxy-setup/README.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ --group //group that the user belongs to
+ --userdel //user that needs to be deleted from the proxy services
+```
+
+An example of use of the parameters can be found from docker compose file
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: s/from/in the

File: cmd/security-proxy-setup/README.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+## Docker Build
+
+Go to the root directory of the repository and use the Makefile to build the docker container image for `security-proxy-setup`:
+
+```sh
+make docker_security_proxy_setup
+```
+
+It should create a docker image with the name `edgexfoundry/docker_security_proxy_setup:<version>-dev` if sucessfully built.
```
Comment (mkbhanda) [inline cmd/security-proxy-setup/README.md]: will these be -dev images once released?

File: cmd/security-secretstore-setup/README.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+The binary supports multiple command line parameters 
+
+```sh
+ --init  //run init procedure for security service
+ --insecureSkipVerify //skip server side SSL verification, mainly for self-signed cert
+ --configfile //use different configuration file other than the default
+ --vaultInterval //time to wait between checking Vault status in seconds
+```
+
+An example of use of the parameters can be found from docker compose file
```
Comment (mkbhanda) [inline cmd/security-secretstore-setup/README.md]: s/An example of use of the parameters can be found from docker compose file/An example of using the parameters can be found in the docker compose file

File: cmd/security-secretstore-setup/README.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+## Docker Build
+
+Go to the root directory of the repository and use the Makefile to build the docker container image for `security-secretstore-setup`:
+
+```sh
+make docker_security_secretstore_setup
+```
+
+It should create a docker image with the name `edgexfoundry/docker_security_secretstore_setup:<version>-dev` if sucessfully built.
```
Comment (mkbhanda) [inline cmd/security-secretstore-setup/README.md]: always -dev image?

File: SECURITY.md (hunk 3 above)
Comment (tingyuz) [reply SECURITY.md]: Fixed.

File: README.md [hunk 15]
```diff
@@ -21,6 +21,23 @@ You can check the status of your running EdgeX services by going to http://local
...
+
+- Security-secrets-setup
+- Security-secretstore-setup
+- Security-proxy-setup
+
+When security features are enabled, there are extra steps required to access the resources of EdgeX.
+
+1. The user needs to create an access token and associate every REST request with the access token
+while sending the request. 
+2. The exported external ports (such as 48080, 48081 etc.) will be inaccessible due to security enhancement. Instead all the REST requests need to go through the proxy, and the proxy will redirect the request to individual microservice on behalf of the user.
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply README.md]: no quite understand what s/  means here?

File: SECURITY.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+Once the EdgeX is up and running, the following steps are required to access the resources of EdgeX.
+
+1. The user needs to create an access token and associate every REST request with the security token
+while sending the request. Use "admin" as groupname below as it is the default enable group in the configuration of proxy.
+
+```sh
+docker-compose run security-proxy-setup --useradd=<account> --group=<groupname>
+```
+
+which will create an access token. The token will be like this: eyJpc3MiOiI5M3V3cmZBc0xzS2Qwd1JnckVFdlRzQloxSmtYOTRRciIsImFjY291bnQiOiJhZG1pbmlzdHJhdG9yIn0
```
Comment (jim-wang-intel) [inline SECURITY.md]:
```suggestion
which will create an access token. One example of an access token will be like: eyJpc3MiOiI5M3V3cmZBc0xzS2Qwd1JnckVFdlRzQloxSmtYOTRRciIsImFjY291bnQiOiJhZG1pbmlzdHJhdG9yIn0
```

File: SECURITY.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
+# Security Components in EdgeX Foundry Services
+
+Starting from Fuji release, EdgeX includes enhanced security features. There are 3 major components that are responsible for security
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md (hunk 4 above)
Comment (jim-wang-intel) [inline SECURITY.md]:
```suggestion
Such request now needs to be converted into a request through `proxy service` and provided `access token`.  One example is like this below:
```

File: SECURITY.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+features: 
+
+- [Security-secrets-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secrets-setup/README.md)
+- [Security-secretstore-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secretstore-setup/README.md)
+- [Security-proxy-setup] (https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-proxy-setup/README.md)
+
+Security-secrets-setup is responsible for creating necessary certificates. Security-secretstore-setup is responsible for initializing the secret store that holds various credentials for EdgeX. Security-proxy-setup is responsible for initializating the environment for proxy for EdgeX to protect all the REST API resources, set up related permissions and authentication methods.
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply SECURITY.md]: Fixed.

File: cmd/security-proxy-setup/README.md (hunk 10 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 11 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 12 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: -dev will be there if we build manually from make.

File: cmd/security-secretstore-setup/README.md (hunk 13 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/README.md]: fixed.

File: cmd/security-secretstore-setup/README.md (hunk 14 above)
Comment (tingyuz) [reply cmd/security-secretstore-setup/README.md]: yes

File: README.md (hunk 5 above)
Comment (tingyuz) [reply README.md]: fixed.

File: SECURITY.md (hunk 2 above)
Comment (jdharms) [reply SECURITY.md]: Consider s/the original request be like this below/such as the following

File: SECURITY.md (hunk 16 above)
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md (hunk 4 above)
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: cmd/security-secretstore-setup/main.go [hunk 19]
```diff
@@ -96,7 +96,7 @@ func main() {
 		func() {
 			tokenFile, err := os.OpenFile(absPath, os.O_CREATE|os.O_RDWR, 0600)
 			if err != nil {
-				secretstore.LoggingClient.Error(fmt.Sprintf("unable to open token file at %s with error: %s", absPath, err.Error()))
+				secretstore.LoggingClient.Error(fmt.Sprintf("unable to open token file at %s", absPath))
```
Comment (jim-wang-intel) [inline cmd/security-secretstore-setup/main.go]: Why this change?

Comment (tingyuz) [reply cmd/security-secretstore-setup/main.go]: Ah this got messed up during rebase.. fixed.

File: SECURITY.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+# Get Started
+
+To get started, fetch the latest docker-compose.yml and start the EdgeX containers:
+
+```sh
+wget https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/releases/fuji/compose-files/docker-compose-fuji-1.1.0.yml
+docker-compose up -d
+```
+
+Once the EdgeX is up and running, the following steps are required to access the resources of EdgeX.
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+1. The user needs to create an access token and associate every REST request with the security token
+while sending the request. Use "admin" as groupname below as it is the default enable group in the configuration of proxy.
```
[the remark this answers is not in the document]
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md (hunk 1 above)
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md (hunk 2 above)
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: SECURITY.md (hunk 6 above)
Comment (tingyuz) [reply SECURITY.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 7 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 8 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 9 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

File: cmd/security-proxy-setup/README.md (hunk 10 above)
Comment (tingyuz) [reply cmd/security-proxy-setup/README.md]: fixed.

**Comment by tingyuz** [comment]
recheck
